### ISSUE TITLE
polishing highlighting feature

### ISSUE THREAD

**Issue by eLiNK2gl** [koreader/kindlepdfviewer#295]
it is possible to highlight a text (press 'n'), which I personally use quite a bit.

1) but after a part of text has been highlighted, the highlight mode is automatically deactivated, so if one intends to highlight another part of text on the same page, you'd need to move the cursor all way down again.
2) highlights cannot be comfortably removed. (You need to move the cursor to your highlight and press delete), Ideally, one should be able to remove them one by one ( or all) like bookmarks.

Please finish the beast :)

P.S.: On a side note, I believe whenever something gets (be it a file, bookmark or a highlight) deleted, it would be more comfortable if the previous (or next item) gets activated. Currently, if you delete a file, the cursor gets placed at the begining of the folder (or the list, as is the case with 'Last read').

**Comment by tigran123** [comment]
I agree with your P.S. comment. However, it is not hard to workaround because Page Next/Prev keys work in the file manager as well and they allow you to quickly jump to the required location. I.e. you don't need to move there in small one-line steps with FW-selector.

**Comment by tigran123** [comment]
On the removal of highlights --- it should be very similar to removal of bookmarks, so leave it with me, I'll have a look.

**Comment by tigran123** [comment]
It (removal of highlights) is actually slightly more complicated than deleting bookmarks, but still should be doable, consider me assigned to this (item 2) on your list).

As for item 1) I can also have a look, but I don't promise.

BtwI have to go somewhere in a few minutes and will be back in the late afternoon, so I may not have a chance to look at any of this until tomorrow.

**Comment by eLiNK2gl** [comment]
@tigran123  That would be great. Thanks.

I suppose it would be pretty straightforward to use highlighting coordinates, so we might at some time make use of the following project -> http://code.google.com/p/kindle-annotations/ (at least, for pdf files)

**Comment by tigran123** [comment]
That program would be wonderful if it was not written in Java and didn't require some massive Java library for PDF manipulation.

The sort of thing you are asking is even beyond the capabilities of Adobe Acrobat Reader but on par with those of Adobe Acrobat Professional. You don't expect one to be able to embed "Acrobat Professional" in a Kindle, do you? :)

**Comment by tigran123** [comment]
@eLiNK2gl You use highlighting quite a bit --- have you not noticed a very serious bug with deleting highlights already present? If there is more than one highlight on a page, say, three highlights, and you delete the second highlight then not only the second but also the third one will be deleted. This is because one cannot use # operator on tables containing nil entries between non-nil ones, and we have the following code handling the pressing of "Del" key in highlight mode:

```
self.highlight[self.pageno][k] = nil                  
              -- remove page entry if empty
              if #self.highlight[self.pageno] == 0 then             
                    self.highlight[self.pageno] = nil                 
              end
```

So, currently, deleting highlights is not just uncomfortable, it does not really work if you have more than one highlight on a page. I will try to fix this also.

**Comment by eLiNK2gl** [comment]
@tigran123  I have deleted one or two highlights for testing purposes only. And as luck has it, I had no more than one highlight on a test page.
Other than that they are almost holy to me. That's why I prefer the deletion of files from history with confirmation only.

**Comment by tigran123** [comment]
Ok, the fixes are in pull request #296 now.

**Comment by tigran123** [comment]
Just to clarify that only the fix for item 2) is in the pull request. The item 1) I did not fix.

**Comment by tigran123** [comment]
@eLiNK2gl  Actually, with my fixes I have broken your "holy of holies", i.e. the highlights. I am sorry about it.
But don't worry, I'll fix them tomorrow! :)

**Comment by eLiNK2gl** [comment]
@tigran123 Oops, I overlooked your question regarding the software provided in the link above. Sorry.

Actually, I do not expect it to be integrated in kpdfviewer.
The purpose of the program is different.
As the native pdf viewer in Kindle cannot embed highlights and annotations to pdf files, the kindle annotator merges the highlights and annotations made in a pdf file to this file on a pc. After all, not everyone has a Kindle.
It is the same with the kpdfviewer.

Regarding highlights, even if the structure of highlights changes it should be relatively easy to make the necessary conversion to the old file. I would be most probably able to do it myself, but not in C or lua.

**Comment by tigran123** [comment]
Btw, I have just now fixed the crash I mentioned above and pushed it my origin tree (not upstream!)
I will test it more thoroughly tomorrow before doing a pull request.
